engine/src/snapshot/logic.rs

```diff
@@ -59,120 +63,120 @@ pub enum WriteError {
-
-    // create plain text buffer.
-    let mut pt = vec![0; ct.len()];
-
-    // decrypt the ciphertext into the plain text buffer.
-    XChaCha20Poly1305::try_decrypt(&shared.to_bytes(), &nonce, associated_data, &mut pt, &ct, &tag)
-        .map_err(|e| ReadError::CorruptedContent(format!("Decryption failed: {}", e)))?;
+/// Decrypt snapshot content with key using maximum work factor recommended for password-based (weak) keys.
+///
+/// Decryption may fail if the required amount of computation (work factor) exceeds the recommended value.
+/// In this case `decrypt_content_with_work_factor` with a larger work factor.
+pub fn decrypt_content<I: Read>(
+    input: &mut I,
+    key: &Key,
+    associated_data: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, ReadError> {
+    let max_work_factor = age::RECOMMENDED_MAXIMUM_DECRYPT_WORK_FACTOR;
+    decrypt_content_with_work_factor(input, key, max_work_factor, associated_data)
+}
 
-    Ok(pt)
+/// Decrypt snapshot content with key using custom maximum work factor.
+///
+/// Decryption may fail if the required amount of computation (work factor) exceeds the provided value.
+/// In this case a larger maximum work factor value can be used.
+///
+/// Strong keys are expected to use small/zero work factor.
+/// Small/zero maximum work factor can be used in such case.
+///
+/// # Security
+///
+/// Key derivation time grows exponentially with work factor.
+/// Maximum work factor should not be too large.
+/// Large values of maximum work factor when exploited by an attacker can cause Denial-of-Service.
+pub fn decrypt_content_with_work_factor<I: Read>(
+    input: &mut I,
+    key: &Key,
+    max_work_factor: u8,
+    _associated_data: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, ReadError> {
+    let mut age = Vec::new();
+    input.read_to_end(&mut age)?;
+
+    age::decrypt_vec(key, max_work_factor, &age[..])
+        .map(Zeroizing::new)
+        .map_err(From::from)
 }
 
-/// Atomically encrypt, add magic and version bytes as file-header, and [`write`][self::write] the specified
+/// Put magic and version bytes as file-header, [`encrypt_content`][self::encrypt_content] the specified
 /// plaintext to the specified path.
 ///
 /// This is achieved by creating a temporary file in the same directory as the specified path (same
 /// filename with a salted suffix). This is currently known to be problematic if the path is a
 /// symlink and/or if the target path resides in a directory without user write permission.
-pub fn write_to(plain: &[u8], path: &Path, key: &Key, associated_data: &[u8]) -> Result<(), WriteError> {
+pub fn encrypt_file(plain: &[u8], path: &Path, key: &Key, associated_data: &[u8]) -> Result<(), WriteError> {
     // TODO: if path exists and is a symlink, resolve it and then append the salt
     // TODO: if the sibling tempfile isn't writeable (e.g. directory permissions), write to
+    if !associated_data.is_empty() {
+        return Err(WriteError::UnsupportedAssociatedData);
+    }
```

Comment:
Why is associated_data even a parameter if it's not supported?

Reply:
Good point. Initially I didn't want to break the API. Now we can get rid of it.

Reply:
Nice, but should also get a changelog entry

Reply:
Added changelog entry.